core/src/main/java/io/questdb/cutlass/line/http/LineHttpSender.java

```diff
@@ -0,0 +1,637 @@
+            sink.putUtf8(chunk.lo(), chunk.hi());
+        }
+    }
+
+    private static boolean isSuccessResponse(DirectUtf8Sequence statusCode) {
+        return statusCode != null && statusCode.size() == 3 && statusCode.byteAt(0) == '2';
+    }
+
+    private static boolean keepAliveDisabled(HttpClient.ResponseHeaders response) {
+        DirectUtf8Sequence connectionHeader = response.getHeader(HttpConstants.HEADER_CONNECTION);
+        return connectionHeader != null && Chars.equals("close", connectionHeader.asAsciiCharSequence());
+    }
+
+    private static long unitToNanos(ChronoUnit unit) {
+        switch (unit) {
+            case NANOS:
+                return 1;
+            case MICROS:
+                return 1_000;
+            case MILLIS:
+                return 1_000_000;
+            case SECONDS:
+                return 1_000_000_000;
+            default:
+                return unit.getDuration().toNanos();
+        }
+    }
+
+    private void consumeChunkedResponse(HttpClient.ResponseHeaders response) {
+        if (!response.isChunked()) {
+            return;
+        }
+        ChunkedResponse chunkedRsp = response.getChunkedResponse();
+        while ((chunkedRsp.recv()) != null) {
+            // we don't care about the response, just consume it so it won't stay in the socket receive buffer
+        }
+    }
+
+    private void escapeQuotedString(CharSequence name) {
+        for (int i = 0, n = name.length(); i < n; i++) {
+            char c = name.charAt(i);
+            switch (c) {
+                case ' ':
+                case ',':
+                case '=':
+                case '\n':
+                case '\r':
+                case '\\':
+                    request.put((byte) '\\').put((byte) c);
+                    break;
+                default:
+                    request.put(c);
+                    break;
+            }
+        }
+    }
+
+    private void escapeString(CharSequence value) {
+        for (int i = 0, n = value.length(); i < n; i++) {
+            char c = value.charAt(i);
+            switch (c) {
+                case '\n':
+                case '\r':
+                case '"':
+                case '\\':
+                    request.put((byte) '\\').put((byte) c);
+                    break;
+                default:
+                    request.put(c);
+                    break;
+            }
+        }
+    }
+
+    private void flush0(boolean closing) {
+        if (state != RequestState.EMPTY && !closing) {
+            throw new LineSenderException("Cannot flush buffer while row is in progress. Use sender.at() or sender.atNow() to finish the current row first.");
+        }
+        if (pendingRows == 0) {
+            return;
+        }
+
+        long retryingDeadlineNanos = Long.MIN_VALUE;
+        int retryBackoff = RETRY_INITIAL_BACKOFF_MS;
+        for (; ; ) {
+            try {
+                HttpClient.ResponseHeaders response = request.send(host, port);
```

Comment:
The only response format we currently support is chunked. So, maybe we should `assert response.isChunked()` somewhere?

Reply:
good idea!

Reply:
actually, I think that's not entirely correct. successful responses are not chunked - because they have no payload at all.

Reply:
Well, we could assert that unsuccessful responses are chunked then.